HeterogeneousCore/CUDAUtilities/interface/HistoContainer.h

```diff
@@ -16,66 +16,79 @@
 #include "HeterogeneousCore/CUDAUtilities/interface/cudastdAlgorithm.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/cuda_assert.h"
-#ifdef __CUDACC__
 #include "HeterogeneousCore/CUDAUtilities/interface/prefixScan.h"
-#endif
 #include "HeterogeneousCore/CUDAUtilities/interface/AtomicPairCounter.h"
 
 
-#ifdef __CUDACC__
 namespace cudautils {
 
   template<typename Histo, typename T>
   __global__
   void countFromVector(Histo * __restrict__ h,  uint32_t nh, T const * __restrict__ v, uint32_t const * __restrict__ offsets) {
-     auto i = blockIdx.x * blockDim.x + threadIdx.x;
-     if(i >= offsets[nh]) return;
-     auto off = cuda_std::upper_bound(offsets, offsets + nh + 1, i);
-     assert((*off) > 0);
-     int32_t ih = off - offsets - 1;
-     assert(ih >= 0);
-     assert(ih < nh);
-     (*h).count(v[i], ih);
+     int first = blockDim.x * blockIdx.x + threadIdx.x;
+     for (int i = first; i < offsets[nh]; i += gridDim.x*blockDim.x) {
+       auto off = cuda_std::upper_bound(offsets, offsets + nh + 1, i);
+       assert((*off) > 0);
+       int32_t ih = off - offsets - 1;
+       assert(ih >= 0);
+       assert(ih < nh);
+       (*h).count(v[i], ih);
+     }
   }
 
   template<typename Histo, typename T>
   __global__
   void fillFromVector(Histo * __restrict__ h,  uint32_t nh, T const * __restrict__ v, uint32_t const * __restrict__ offsets) {
-     auto i = blockIdx.x * blockDim.x + threadIdx.x;
-     if(i >= offsets[nh]) return;
-     auto off = cuda_std::upper_bound(offsets, offsets + nh + 1, i);
-     assert((*off) > 0);
-     int32_t ih = off - offsets - 1;
-     assert(ih >= 0);
-     assert(ih < nh);
-     (*h).fill(v[i], i, ih);
+     int first = blockDim.x * blockIdx.x + threadIdx.x;
+     for (int i = first; i < offsets[nh]; i += gridDim.x*blockDim.x) {
+       auto off = cuda_std::upper_bound(offsets, offsets + nh + 1, i);
+       assert((*off) > 0);
+       int32_t ih = off - offsets - 1;
+       assert(ih >= 0);
+       assert(ih < nh);
+       (*h).fill(v[i], i, ih);
+     }
   }
 
   template<typename Histo>
-  void launchZero(Histo * __restrict__ h, cudaStream_t stream) {
+  void launchZero(Histo * __restrict__ h, cudaStream_t stream = 0) {
```

Comment:
Downside is that now on CUDA side one is allowed to e.g. forget the `stream` in which case the kernel goes to the default stream.